Add hybrid search blog (#2182)
opensearch-project/project-website: _posts/2023-09-21-hybrid-search.md

---
layout: post
title:  Hybrid search is generally available in OpenSearch 2.10

> authors:

Reviewer: I think we need to add Vamshi as well, @navneet1v what do you think?

Reviewer 2: Yes lets add him too

## Next steps

We have identified several areas of improvement for hybrid search, and we’re planning to work on them for the future OpenSearch releases. In the short term, a good starting point is to improve performance by running individual queries of the main hybrid query in parallel instead of sequentially. This should significantly improve latency, especially when all inner queries have similar running times. We are considering the following improvements for the future releases:

* Executing individual queries in parallel.
* Adding more configuration options and parameters to the normalization processor to allow more control over combined results. For instance, we can add the ability to specify a minimal score for documents to be returned in the results, which will avoid returning non-competitive hits. 
* Supporting results pagination
* Supporting filters in the hybrid query clause. It’s possible to define a filter for each inner query individually, but it’s not optimal if a filter condition is the same for all inner queries.
* Adding more benchmark results for larger datasets so we can provide recommendations for using hybrid search in various configurations. 

> 

Reviewer 2: we have github issues around all of them. @martin-gaievski please provide those issues here

Reviewer: sure, here are the issues I opened for future work:

- https://github.com/opensearch-project/neural-search/issues/279
- https://github.com/opensearch-project/neural-search/issues/280
- https://github.com/opensearch-project/neural-search/issues/281
- https://github.com/opensearch-project/neural-search/issues/282
- https://github.com/opensearch-project/neural-search/issues/299

---
layout: post
title:  Improve search relevance with hybrid search, generally available in OpenSearch 2.10
authors:
  - gaievski
  - navneev
  - vamshin
  - kolchfa
date: 2023-09-21
categories:
  - technical-posts

> meta_keywords: 

Reviewer 3: Please update with the following meta: 

**meta_keywords:**  Improve search relevance, hybrid search in OpenSearch 2.10, semantic and keyword search, 
**meta_description:**  Improve search relevance with OpenSearch 2.10 when you tune search relevance by using hybrid search to combine and normalize query relevance scores.

        <td>83.2</td>
        <td>1.9</td>
    </tr>
    <tr>
        <td>Quora</td>
        <td>61</td>
        <td>69</td>
        <td>8</td>
        <td>69</td>
        <td>78</td>
        <td>9</td>
        <td>73.4</td>
        <td>84</td>
        <td>10.6</td>
    </tr>
    <tr>
        <td>Amazon ESCI</td>
        <td>49</td>
        <td>50</td>
        <td>1</td>
        <td>58</td>
        <td>59.4</td>
        <td>1.4</td>
        <td>67</td>
        <td>70</td>
        <td>3</td>
    </tr>
    <tr>
        <td>DBPedia</td>
        <td>100.8</td>
        <td>107.7</td>
        <td>6.9</td>
        <td>117</td>
        <td>130.9</td>
        <td>13.9</td>
        <td>129.8</td>
        <td>150.2</td>
        <td>20.4</td>
    </tr>
    <tr>
        <td>FiQA</td>
        <td>53.9</td>
        <td>56.9</td>
        <td>3</td>
        <td>61.9</td>
        <td>65</td>
        <td>3.1</td>
        <td>64</td>
        <td>67.7</td>
        <td>3.7</td>
    </tr>
    <tr>
        <td><b>% change vs. Boolean query </b></td>
        <td colspan="3"><b>6.40%</b></td>
        <td colspan="3"><b>6.96%</b></td>
        <td colspan="3"><b>8.27%</b></td>
    </tr>
</table>

> As shown in the preceding table, hybrid search improves the result quality by 8&ndash;12% compared to keyword search and by 15% compared to natural language search. Simultaneously, a hybrid query exhibits a 6&ndash;8% increase in latency compared to a Boolean query when executing the same inner queries. 

Author: Leaving as is because of 6

## Conclusions

In this blog, we have included series of experiments that show that hybrid search produces results that are very close to theoretical expectations and in most cases are better than the results of individual queries alone. Hybrid search comes with a certain increase in latency, which will be addressed in future versions (you can track the progress of this enhancement in [this GitHub issue](https://github.com/opensearch-project/neural-search/issues/279)).

It’s important to remember that datasets have different parameters and for some datasets hybrid search may not consistently improve results. Conversely, experimenting with different parameters&mdash;for instance, selecting higher values of `k` or a different `space_type` for neural search&mdash;may lead to better results. 

We’ve seen that some of the conclusions may be applied to most datasets:

* For semantic search, hybrid query with normalization produces better results compared to neural search or text search alone.
* The combination of `min-max` score normalization and `arithmetic_mean` score combination achieves the best results, compared to other techniques.

> * In most cases, increasing the value of `k` in k-NN data type leads to better results up to a certain point, but after that there is no increase in relevance. At the same time, high values of `k` increase search latency, so from our observations, it’s better to choose the value of `k` between 100 and 200. 

Reviewer 4: ```suggestion
* In most cases, increasing the value of `k` in the k-NN data type leads to better results up to a certain point, but after that, there is no increase in relevance. At the same time, high values of `k` increase search latency, so from our observations, it’s better to choose a value of `k` between 100 and 200. 
```

> * We have seen best results when `innerproduct` is specified as a space type for k-NN vector fields. This may be because our models were trained using the inner product similarity function.

Reviewer 4: ```suggestion
* The best results are produced when `innerproduct` is specified as a space type for k-NN vector fields. This may be because our models were trained using the inner product similarity function.
```

> * An increase in search relevance comes with a cost of a 6&ndash;8% increase in latency, which should be acceptable in most cases.

Author: Changed to "leads to"

> In general, as our experiments demonstrated, hybrid search produces results that are very close to the ones described by the science team, so [all their conclusions](https://opensearch.org/blog/semantic-science-benchmarks/#section-5-strengths-and-limitations) are applicable to hybrid search. 

Author: I've included the link to the original blog post, not just the results. @martin-gaievski  Do you have any other suggestions fro clarifying this?

## Next steps

We have identified several areas of improvement for hybrid search, and we’re planning to work on them for the future OpenSearch releases. In the short term, a good starting point is to improve performance by running individual queries of the main hybrid query in parallel instead of sequentially. This should significantly improve latency, especially when all inner queries have similar running times. 

> We are considering the following improvements for the future releases:

Reviewer 4: ```suggestion
We are considering including the following improvements in future versions:
```

* Executing individual queries in parallel (see [issue #279](https://github.com/opensearch-project/neural-search/issues/279) and [issue #281](https://github.com/opensearch-project/neural-search/issues/281)).
* Adding more configuration options and parameters to the normalization processor to allow more control over combined results. For instance, we can add the ability to specify a minimal score for documents to be returned in the results, which will avoid returning non-competitive hits (see [issue #299](https://github.com/opensearch-project/neural-search/issues/299)). 
* Supporting results pagination (see [issue #280](https://github.com/opensearch-project/neural-search/issues/280)).

> * Supporting filters in the hybrid query clause (see [issue #282](https://github.com/opensearch-project/neural-search/issues/281)). It’s possible to define a filter for each inner query individually, but it’s not optimal if a filter condition is the same for all inner queries.

Reviewer 4: ```suggestion
* Supporting filters in the hybrid query clause (see [issue #282](https://github.com/opensearch-project/neural-search/issues/281)). It’s possible to define a filter for each inner query individually, but it’s not optimal for a filter condition to be the same for all inner queries.
```

* Adding more benchmark results for larger datasets so we can provide recommendations on using hybrid search in various configurations. 

## Dataset statistics

The following table provides further details of the test datasets used for benchmarking. 

> |Dataset	|Average query length	|Average query length	|Average query length	|Average query length	|Average query length	|Average query length	|

Reviewer 2: The headings are all same. :( This needs to be fixed.

Reviewer 2: @krisfreedain , @kolchfa-aws
